uqcsbot/scripts/weather.py

```diff
@@ -0,0 +1,214 @@
+    precip_prob = node.find(".//text[@type='probability_of_precipitation']")
+    if rain_range is not None and precip_prob is not None:
+        return "{} Chance of Precipitation; {}".format(precip_prob.text, rain_range.text)
+    elif precip_prob is not None:
+        return "{} Chance of Precipitation".format(precip_prob.text)
+    return ""
+
+
+def response_brisbane_detailed():
+    """
+    returns a detailed forecast for Brisbane"
+    """
+    try:
+        data = urlopen("ftp://ftp.bom.gov.au/anon/gen/fwo/IDQ10605.xml")
+        root = ET.fromstring(data.read())
+    except Exception:
+        return ""
+    node = root.find(".//area[@description='Brisbane']")
+    if node is None:
+        return ""
+    node = node.find(".//forecast-period[@index='0']")
+    if node is None:
+        return ""
+
+    forecast = node.find(".//text[@type='forecast']")
+    forecast = "" if forecast is None else forecast.text
+
+    fire_danger = node.find(".//text[@type='fire_danger']")
+    if fire_danger is None or fire_danger.text == "Low-Moderate":
+        fire_danger = ""
+    else:
+        fire_danger = "There Is A {} Fire Danger Today".format(fire_danger.text)
+
+    uv_alert = node.find(".//text[@type='uv_alert']")
+    uv_alert = "" if uv_alert is None else uv_alert.text
+
+    return (forecast, fire_danger, uv_alert)
+
+
+@bot.on_command('weather')
+def handle_weather(command: Command):
+    """
+    `!weather [[state] location] [day]` - Returns the weather forcaset for a location
+    `day` is how many days into the future the forecast is for (0 is today and default)
+    `location` defaults to Brisbane, and `state` defualts to QLD
+    """
+
+    (state, location, future) = process_arguments(command.arg)
+
+    root = get_xml(state)
+    if root is None:
+        bot.post_message(command.channel_id, "Could Not Retrieve BOM Data")
+        return
+
+    node, response = find_location(root, location, future)
+    if node is None:
+        bot.post_message(command.channel_id, response)
+        return
+
+    # get responses
+    response = []
+    response.append(response_header(node, location))
+    response.append(response_overall(node))
+    response.append(response_temperature(node))
+    response.append(response_precipitation(node))
+    # post
+    bot.post_message(command.channel_id, "\r\n".join([r for r in response if r]))
+
+
+@bot.on_schedule('cron', hour=6, minute=0, timezone='Australia/Brisbane')
+def daily_weather():
+    """
+    Posts today's Brisbane weather at 6:00am every day
+    """
+
+    (state, location, future) = ("QLD", "Brisbane", 0)
+
+    root = get_xml(state)
+    if root is None:
+        return
+
+    node, response = find_location(root, location, future)
+    if node is None:
+        return
+
+    # get responses
+    response = []
+    brisbane_detailed, brisbane_fire, brisbane_uv = response_brisbane_detailed()
+    response.append(response_header(node, location))
```

Comment:
Could all of this be in a single fstring? If not it could certainly use a helper method.

Reply:
Possibly, but it would likely look inelegant, especially as some fields don't exist depending on time of day and days into the future the forecast is for etc.